### ISSUE TITLE
SFN.Client.exceptions.ExecutionAlreadyExists Not implemented

### ISSUE THREAD

**Issue by ciaranevans** [spulec/moto#3262]
**TL;DR: When calling `start_execution`, if a execution exists already with the provided name, throw `SFN.Client.exceptions.ExecutionAlreadyExists`**

## Expected behaviour:

When using a `mock_stepfunctions` like so:

```python
import boto3
from moto import mock_stepfunctions

def get_mock_step_function_arn(step_function_client):
    """
    Create a single node Step Function that passes
    """
    step_function = step_function_client.create_state_machine(
        name="TestStepFunction",
        roleArn="arn:aws:iam::123456789012:role/TestRole",
        definition=json.dumps(
            {
                "StartAt": "Pass",
                "States": {"Pass": {"Type": "Pass", "Result": "Blah", "End": True}},
            }
        ),
    )
    return step_function["stateMachineArn"]

@mock_stepfunctions
def test_that_duplicate_execution_fails():
    step_function_client = boto3.client('stepfunctions')
    mock_step_function_arn = get_mock_step_function_arn(step_function_client)
    
    step_function_client.start_execution(stateMachineArn=mock_step_function_arn, name="a-name", input="")
    # Below should throw SFN.Client.exceptions.ExecutionAlreadyExists
    step_function_client.start_execution(stateMachineArn=mock_step_function_arn, name="a-name", input="")
```

I expect the second call to `start_execution` to fail, throwing `SFN.Client.exceptions.ExecutionAlreadyExists` as the name of the execution has already been used.

## Actual behaviour:

The actual behaviour is the execution happens, and a `list_executions` call returns both the original and the duplicate.

## Versions

**moto** `1.3.14`
**boto** `2.49.0`
**boto3** `1.14.45`
**botocore** `1.17.45`

All were installed via `pip` and I'm *not* using the standalone `moto`

**Comment by ciaranevans** [comment]
I'd be interested in contributing a fix to this, but don't want to dive in without bouncing ideas off folks first!

I'm thinking https://github.com/spulec/moto/blob/c0953a229f959de79d22f3ed60f273eaa1800fcf/moto/stepfunctions/models.py#L206-L216
would have something like
```python
 def _execution_exists(self, name)
     for execution in self.executions:
         if execution.name == name:
             return True
     return False

 def start_execution(self, state_machine_arn, name=None): 
     state_machine_name = self.describe_state_machine(state_machine_arn).name
     if _execution_exists(name):
         raise ExecutionAlreadyExists("<message-like-AWS-would-provide>")
     execution = Execution( 
         region_name=self.region_name, 
         account_id=self._get_account_id(), 
         state_machine_name=state_machine_name, 
         execution_name=name or str(uuid4()), 
         state_machine_arn=state_machine_arn, 
     ) 
     self.executions.append(execution) 
     return execution 
```

The above is super high level, but I've poked around the Exceptions etc to see where stuff lives!

**Comment by ciaranevans** [comment]
@bblommers I've just installed `1.3.15.dev1032` but now all tests using `moto` are failing due to InvalidAccessKeyId etc.

It looks like a change has been made in how credentials are handled somewhere down the line?

For reference, beforehand I was using `1.3.14`

**Comment by bblommers** [comment]
The handling of credentials did change @ciaranevans - see the readme on some guidance how to fix this: https://github.com/spulec/moto#very-important----recommended-usage

**Comment by ciaranevans** [comment]
Ah, I think what we're seeing is the imports issue. 

That's unfortunate for unit tests!

**Comment by ciaranevans** [comment]
@bblommers I've now done imports of the to be testing methods into their test methods and can confirm that the ExecutionAlreadyExists exception is now being thrown correctly.

Going to call this issue sorted!

If you've got a nice way to handle having to import something every test function I write, I'd love to hear it!

**Comment by chriselion** [comment]
Just for posterity, I think this check was too aggressive; it should allow the same name as long as the input is the same. I created https://github.com/getmoto/moto/issues/9394 as a followup.
